src/Pate/Verification/PairGraph.hs

```diff
@@ -0,0 +1,878 @@
+       Widen _ d'' ->
+         case updateDomain gr currBlock pPair d'' of
+           Nothing ->
+             do traceBundle bundle ("Ran out of gas while widening postconditon! " ++ show currBlock ++ " " ++ show pPair)
+                return gr -- TODO? better error handling?
+           Just gr' ->
+             do traceBundle bundle "Successfully widened postcondition"
+                return gr'
+
+getTargetDomain ::
+  PPa.BlockPair arch ->
+  PairGraph sym arch ->
+  EquivM sym arch (AbstractDomain sym arch)
+getTargetDomain pPair gr =
+  case Map.lookup pPair (pairGraphDomains gr) of
+    Just d  -> return d
+    Nothing ->
+      -- initial state of the pair graph: choose the universal domain that equates as much as possible
+      PVD.universalDomainSpec pPair
+
+data WidenLocs sym arch =
+  WidenLocs
+    (Set (Some (MM.ArchReg arch)))
+    (Set (Some (PMc.MemCell sym arch)))
+
+instance PA.ValidArch arch => Show (WidenLocs sym arch) where
+  show (WidenLocs regs cells) =
+    unlines [ unwords (map show (Set.toList regs))
+            , show (Set.size cells) ++ " memory locations"
+            ]
+
+instance (OrdF (W4.SymExpr sym), PA.ValidArch arch) => Semigroup (WidenLocs sym arch) where
+  (WidenLocs r1 m1) <> (WidenLocs r2 m2) = WidenLocs (r1 <> r2) (m1 <> m2)
+
+instance (OrdF (W4.SymExpr sym), PA.ValidArch arch) => Monoid (WidenLocs sym arch) where
+  mempty = WidenLocs mempty mempty
+
+data WidenResult sym arch
+  = NoWideningRequired
+  | WideningError String
+  | Widen (WidenLocs sym arch) (AbstractDomain sym arch)
+
+-- | Try the given widening stratetigs s one at a time,
+--   until the first one that computes some nontrival
+--   widening, or returns an error.
+tryWidenings ::
+  [IO (WidenResult sym arch)] ->
+  IO (WidenResult sym arch)
+tryWidenings [] = return NoWideningRequired
+tryWidenings (x:xs) =
+  x >>= \case
+    NoWideningRequired -> tryWidenings xs
+    res -> return res
+
+widenPostcondition :: forall sym arch.
+  PA.ValidArch arch =>
+  SimBundle sym arch ->
+  AbstractDomain sym arch {- ^ predomain -} ->
+  AbstractDomain sym arch {- ^ postdomain -} ->
+  EquivM sym arch (WidenResult sym arch)
+widenPostcondition bundle preD postD0 =
+  withSym $ \sym ->
+    do vcfg <- asks envConfig
+       asmFrame <- asks envCurrentFrame
+       eqRel <- asks envBaseEquiv
+       stackRegion <- asks (PMC.stackRegion . envCtx)
+
+       let solver = PCfg.cfgSolver vcfg
+       let saveInteraction = PCfg.cfgSolverInteractionFile vcfg
+
+       precond <- liftIO $ do
+         asm <- PS.getAssumedPred sym asmFrame
+         eqInputs <- PE.getPredomain sym stackRegion bundle eqRel (PS.specBody preD)
+         W4.andPred sym asm eqInputs
+
+       -- traceBundle bundle "== widenPost: precondition =="
+       -- traceBundle bundle (show (W4.printSymExpr precond))
+
+       PS.withOnlineSolver solver saveInteraction sym $ \bak ->
+         do liftIO $ LCBO.withSolverProcess bak doPanic $ \sp -> do
```

Comment:
Is this also the kind of thing that needs a new frame? Or is there a higher-level protocol maintained in this loop that means that we are already under a frame?

Reply:
Same situation as above, basically.  This is a fresh solver, and the slice preconditions need to be in scope for everything we are doing below.